pkg/autoupdate/tuf/library_manager.go

```diff
@@ -0,0 +1,341 @@
+type updateLibraryManager struct {
+	metadataClient  *client.Client // used to validate downloads
+	mirrorUrl       string         // dl.kolide.co
+	mirrorClient    *http.Client
+	rootDirectory   string
+	operatingSystem string
+	osquerier       localserver.Querier // used to query for current running osquery version
+	logger          log.Logger
+}
+
+func newUpdateLibraryManager(metadataClient *client.Client, mirrorUrl string, mirrorClient *http.Client, rootDirectory string, operatingSystem string, osquerier localserver.Querier, logger log.Logger) (*updateLibraryManager, error) {
+	ulm := updateLibraryManager{
+		metadataClient:  metadataClient,
+		mirrorUrl:       mirrorUrl,
+		mirrorClient:    mirrorClient,
+		rootDirectory:   rootDirectory,
+		operatingSystem: operatingSystem,
+		osquerier:       osquerier,
+		logger:          log.With(logger, "component", "tuf_autoupdater_library_manager"),
+	}
+
+	// Ensure our staged updates and updates directories exist
+	for _, binary := range binaries {
+		// Attempt out the staged updates directory, in case there are any archives there that we
+		// were unable to clean up previously.
+		_ = os.RemoveAll(ulm.stagedUpdatesDirectory(binary))
+
+		// Create the directory for staging updates
+		if err := os.MkdirAll(ulm.stagedUpdatesDirectory(binary), 0755); err != nil {
+			return nil, fmt.Errorf("could not make staged updates directory for %s: %w", binary, err)
+		}
+
+		// Create the update library
+		if err := os.MkdirAll(ulm.updatesDirectory(binary), 0755); err != nil {
+			return nil, fmt.Errorf("could not make updates directory for %s: %w", binary, err)
+		}
+	}
+
+	return &ulm, nil
+}
+
+// updatesDirectory returns the update library location for the given binary.
+func (ulm *updateLibraryManager) updatesDirectory(binary string) string {
+	return filepath.Join(ulm.rootDirectory, fmt.Sprintf("%s-updates", binary))
+}
+
+// stagedUpdatesDirectory returns the location for staged updates -- i.e. updates
+// that have been downloaded, but not yet verified.
+func (ulm *updateLibraryManager) stagedUpdatesDirectory(binary string) string {
+	return filepath.Join(ulm.rootDirectory, fmt.Sprintf("%s-staged-updates", binary))
+}
+
+// addToLibrary adds the given target file to the library for the given binary,
+// downloading and verifying it if it's not already there. After any addition
+// to the library, it cleans up older versions that are no longer needed.
+func (ulm *updateLibraryManager) AddToLibrary(binary string, targetFilename string) error {
+	// Check to see if the current running version is the version we were requested to add;
+	// return early if it is, but don't error out if we can't determine the current version.
+	currentVersion, err := ulm.currentRunningVersion(binary)
+	if err != nil {
+		level.Debug(ulm.logger).Log("msg", "could not get current running version", "binary", binary, "err", err)
+	} else {
+		if currentVersion.Original() == ulm.versionFromTarget(binary, targetFilename) {
```

Comment:
This makes sense, but `currentVersion.Original() == ulm.versionFromTarget(binary, targetFilename)` feels awkward. Maybe the simplest though

Reply:
I agree it's a little awkward -- I'd looked into having `ulm.versionFromTarget` parse and return a `*semver.Version` but most of the places consuming that function just needed the string, so it didn't feel worth it.